### PR TITLE
Eth1 withdrawal credentials (0x01)

### PR DESCRIPTION
**Pull request by djrtwo**
Add eth1 withdrawal credentials by specifying a new withdrawal prefix version (0x01) whose content is a 20-byte eth1 address.

Note, this specifies the minimum required commitment and leaves out the end-to-end details on how the beacon chain and eth1 chain/state will handle the withdrawals other than the fact that the ETH will be transferred into the specified address with no payload other than the ETH and this will be triggered/payed for through a normal eth1 TX.

See this [ethresearch post](https://ethresear.ch/t/simple-eth1-withdrawals-beacon-chain-centric/8256) for an end-to-end *version* of how withdrawals might take place. Note, the only important details for this commitment are that 0x01 withdrawals receipts automatically make it to the eth1 chain and then can be consumed through a normal TX that specifies gas/tx-fees and sends the ETH to the destination address.

This opens up eth1 custody options such as smart contract wallets and simple hardware wallet support along with enabling a number of trustless staking pool options

-------------

Please let me know if the amount specified in `validator.md` is clear enough.

I considered creating a fork agnostic folder `specs/withdrawal_credentials` to specify each withdrawal_credential version independent of any fork. Given that these can't really change once deployed, it might make sense to gather them all in one place. Any thoughts?

### CODE REVIEW

File: specs/phase0/validator.md
Comment (dankrad) [inline specs/phase0/validator.md:131]: I assume this means that a classic Eth1 account will have to make a call to a precompile that executes the withdrawal? And that Eth1 account will pay for the gas?

Comment (djrtwo) [reply specs/phase0/validator.md:131]: Yes

The idea is that eth1 is aware of receipts and that a classic eth1 TX can trigger the consumption of a receipt as a send to the destination address. Upon success, the receipt is confirmed as consumed

I'm hesitant to go deep into the details here because they aren't set in stone, but the general "eth1 TX pays for gas to trigger receipt consumption and ETH transfer" is the commitment being made in the PR.  That said, I agree what I've written might not be abundantly clear. *Open to suggestions to clarify*

**Comment by dankrad** [comment]
Generally speaking I would like to have this in the spec. I think an important function for any Eth1 withdrawal contract is to initiate the validator exit. This is an essential function for any staking pool (where it's not sufficient to use the withdrawal key to initiate exit).

One way to implement it would be to add a function to the withdrawal contract
```
def isExitable(validatorPubkey) -> bool:
```
and any Eth1 address can initiate the "Eth2 exit" precompile that will then use this function to check if the contract allows an exit.

File: tests/core/pyspec/eth2spec/test/phase0/block_processing/test_process_deposit.py
Comment (lsankar4033) [inline tests/core/pyspec/eth2spec/test/phase0/block_processing/test_process_deposit.py:121]: what's the purpose of this test?

File: specs/phase0/validator.md
Comment (lsankar4033) [reply specs/phase0/validator.md:131]: is there any reason to constrain the eth1 account making that call to the withdrawal address? if not, it seems fine to leave this open ended

File: specs/phase0/validator.md
Comment (ajsutton) [inline specs/phase0/validator.md:128]: nit:
```suggestion
Eth1 address credentials are controlled by an eth1 address. This can be either an externally owned account or a contract.
```

**Review by Grubson33421 — APPROVED** [review]

**Comment by zilm13** [comment]
I would add that in case of using `ETH1_ADDRESS_WITHDRAWAL_PREFIX`, Ether will be transferred from fixed `SYSTEM_CONTRACT` address, with exact address added later, so contracts could easy clarify who is the sender.

**Comment by dapplion** [comment]
Should it be part of the spec what minimal functionality should a contract have if it's an eth1 withdrawal address? i.e. be able to send a simple transaction to a TBD address, receive funds, etc

File: tests/core/pyspec/eth2spec/test/phase0/block_processing/test_process_deposit.py
Comment (djrtwo) [reply tests/core/pyspec/eth2spec/test/phase0/block_processing/test_process_deposit.py:121]: to ensure clients can handle malformed withdrawal credentials

**Comment by ralexstokes** [comment]
> I would add that in case of using `ETH1_ADDRESS_WITHDRAWAL_PREFIX`, Ether will be transferred from fixed `SYSTEM_CONTRACT` address, with exact address added later, so contracts could easy clarify who is the sender.

i'd think specifying a mechanism for the implementation of moving ETH from the beacon chain to the evm userspace is a little premature given the unknowns around the details of the eth1-eth2 merge. (e.g. we could imagine tracking withdrawals outside an actual contract in the EVM)

my understanding of this PR is more to provide a firm(er) commitment to eth withdrawals to today's eth1 so that the use cases that can get away with *just* this functionality can move ahead w/ development and implementation.

i do think we eventually want a more involved mechanism for eth transfer (e.g. https://github.com/ethereum/eth2.0-specs/issues/2151) but i'd suggest waiting for a clearer understanding of the merge before moving ahead w/ concrete specifics of such a scheme

**Comment by ralexstokes** [comment]
> Should it be part of the spec what minimal functionality must a contract have if it's an eth1 withdrawal address? i.e. be able to send a simple transaction to a TBD address, receive funds, etc.

The PR does specify that the eth1 withdrawal address just needs to be able to accept ETH (e.g. by default as an EOA, or a `payable` function in a Solidity smart contract). If you look at my prior comment, there is certainly a space for a more complex mechanism for withdrawals (discussion starting here: https://github.com/ethereum/eth2.0-specs/issues/2151) but I'd say that is a another iteration of withdrawal handling once we better understand the eth1-eth2 merge.

**Review by samueldashadrach — APPROVED** [review]
awesome

**Review by samueldashadrach — APPROVED** [review]

**Review by samueldashadrach — APPROVED** [review]
moon it

**Review by lsankar4033 — APPROVED** [review]

**Comment by lsankar4033** [comment]
it seems that this PR is partially held up b/c of a question around whether eth1-triggered withdrawals are necessary for staking pool functionality.

I suspect that the `0x01` creds here are sufficient for some designs and should probably be set in stone now. Can always add other credential types later. what do folks think?

File: specs/phase0/validator.md
Comment (lsankar4033) [inline specs/phase0/validator.md:135]: technically an `Eth1+Eth2` transaction, not an `Eth1` transaction

Comment (djrtwo) [reply specs/phase0/validator.md:135]: ehhh 

maybe "by a user transaction"

Comment (lsankar4033) [reply specs/phase0/validator.md:135]: ya that's better. i think it's important to make clear that this isn't do-able until post-merge

**Comment by lsankar4033** [comment]
👍

File: specs/phase0/validator.md
Comment (djrtwo) [reply specs/phase0/validator.md]: done!

**Comment by djrtwo** [comment]
@zilm13 

> Ether will be transferred from fixed SYSTEM_CONTRACT address, with exact address added later, so contracts could easy clarify who is the sender.

I'm inclined to not even put this in there as it is not an entirely actionable detail today.
Also, assume this is a contract that can split funds or a multi-sig or something. Receiving extra ETH from non `SYSTEM_CONTRACT` seems fine in almost all cases 😂

**Review by Rose2161 — APPROVED** [review]

**Comment by AmethystWizard** [comment]
Unstaking BLS to ETH1 - the wagyu gui asks for a 24 word recovery phrase, but metamask offers a 12 word phrase, how to proceed in that case?

**Comment by ralexstokes** [comment]
I don't know how the wagyu GUI works and you should probably ask the maintainers

You may find more information here: https://launchpad.ethereum.org/en/withdrawals

**Comment by AmethystWizard** [comment]
allnodes.com created a Ledger Nano X compatible tool that ended up working. 
https://help.allnodes.com/en/articles/7235575-how-to-change-your-ethereum-withdrawal-credentials-from-0x00-to-0x01-deposit-cli

**Comment by Romansharar** [comment]
> Add eth1 withdrawal credentials by specifying a new withdrawal prefix version (0x01) whose content is a 20-byte eth1 address.
> 
> Note, this specifies the minimum required commitment and leaves out the end-to-end details on how the beacon chain and eth1 chain/state will handle the withdrawals other than the fact that the ETH will be transferred into the specified address with no payload other than the ETH and this will be triggered/payed for through a normal eth1 TX.
> 
> See this [ethresearch post](https://ethresear.ch/t/simple-eth1-withdrawals-beacon-chain-centric/8256) for an end-to-end _version_ of how withdrawals might take place. Note, the only important details for this commitment are that 0x01 withdrawals receipts automatically make it to the eth1 chain and then can be consumed through a normal TX that specifies gas/tx-fees and sends the ETH to the destination address.
> 
> This opens up eth1 custody options such as smart contract wallets and simple hardware wallet support along with enabling a number of trustless staking pool options
> 
> Please let me know if the amount specified in `validator.md` is clear enough.
> 
> I considered creating a fork agnostic folder `specs/withdrawal_credentials` to specify each withdrawal_credential version independent of any fork. Given that these can't really change once deployed, it might make sense to gather them all in one place. Any thoughts?

i need eth

**Comment by Grossbeatz** [comment]
0x8e790A09F32389F36692bF8396Ef8CB7d7291A0C

**Review by dustpan999 — COMMENTED** [review]
Sorry if this took me a while

**Pull request merged**
